doc/CodeReview.md

```diff
@@ -172,7 +182,190 @@ A meta target that runs every piece of CI possible.
+- Is any new functionality both publicly exported and harmful to overall safety
+  (not just correctness) if called at the wrong time? If so, this should likely
+  be guarded with a capability.
+- Uses of `#inline` directives should be explained why they are needed.
+
+### Review Guide by Repository Subsystem
+
+In addition to general code review practices, certain review principles are only
+applicable in specific portions of the tock kernel repository.
+
+#### Core Kernel (`/kernel` crate) Not Including HILs
+
+In general, any substantial changes to the kernel crate should be accompanied
+first by a discussion issue. This permits discussion on if the change should be
+included in the Tock kernel and if so how it should be implemented.
+
+All substantial changes should be clearly documented. New files should use `//!`
+comments to explain their purpose, and all functions and data structures should
+be clearly documented with `///` comments. Often additional documentation in
+discrete markdown files is required as well.
+
+Additionally, particularly subtle or extensively discussed rationale should be
+included in the source file directly (often with a `//` comment). This leaves a
+clear trace of how key design decisions in Tock were decided and why certain
+aspects may not use the most intuitive design. This helps avoid re-hashing
+discussions and assist new users with understanding the kernel.
+
+All `unsafe` usage MUST be accompanied by a comment starting with `### Safety`
+that discusses exactly why the unsafe code is necessary and what checks are
+needed and completed to ensure the use of `unsafe` does not trigger undefined
+behavior.
+
+All new exports from the core kernel crate must be carefully examined. Certain
+functionality is only safe within the core kernel. As essentially every crate in
+Tock uses `kernel` as a dependency, anything exported can be used broadly.
+Functionality which is sensitive but _must_ be exported must be guarded by a
+capability.
+
+#### HILs
+
+New HILs should follow the TRD on HIL design.
+
+HILs should be well documented and not specifically matched to a single hardware
+platform.
+
+All valid errors should be enumerated.
+
+HIL naming should be reasonably consistent and clear.
+
+#### Capsules
+
+Capsules should explain what they do in comments but do not need to be
+rigorously commented.
+
+**Virtualizers**
+
+Virtualizers multiplex an underlying resource for multiple users.
+
+- The `Mux` struct should handle all interrupts, and route callback to specific
+  virtualizer users.
+- The virtualizer should provide the same interface (i.e. HIL) as it uses from
+  the underlying shared resource.
+
+**Syscall Drivers**
+
+Syscall drivers implement `SyscallDriver` to provide interfaces for userspace.
+
+- These drivers must support multiple processes. They do not need to be fully
+  virtualized, but they must not break if multiple processes use them.
```

Comment:
I think this needs more clarity: Is it okay for them to be first-come-first-served and reject later processes? (but not crash or do something invalid)

Reply:
Yes that is what that was trying to say.